Feat: Add Passbolt Provider

thorbenbelow opened the pull request:
## Problem Statement

Add basic provider implementation for [Passbolt](https://www.passbolt.com/).

## Related Issue

Fixes:
- https://github.com/external-secrets/external-secrets/issues/3074
- https://github.com/external-secrets/external-secrets/issues/2188

## Proposed Changes

Added support for following APIs:
- [x] SecretStore
- [x] ClusterSecretStore
- [x] ExternalSecret with `remoteRef.key`
- [x] ExternalSecret with `dataFrom.find.name.regexp`

### SecretStore and ClusterSecretStore
The passbolt provider requires auth in the form of a password and a private key as well as a hostname to connect to.

```yaml
provider:
    passbolt:
      host: https://passbolt.passbolt.svc.cluster.local
      auth:
        passwordSecretRef:
          key: password
          name: passbolt-credentials
        privateKeySecretRef:
          key: privateKey
          name: passbolt-credentials
```
### External Secret
By default secrets include `name`, `username`, `uri`, `password` and `description` and are in json format.
```
'{"name":"passbolt-secret","username":"some-username","password":"supersecretpassword","uri":"passbolt.com","description":"some description"}'
```
Using the `property` key this can be reduced to only display one single property.

The key for secrets synced with `dataFrom.find.name.regexp` is the respective Passbolt resource ID.

## Checklist

- [x] I have read the [contribution guidelines](https://external-secrets.io/latest/contributing/process/#submitting-a-pull-request)
- [x] All commits are signed with `git commit --signoff`
- [x] My changes have reasonable test coverage
- [x] All tests pass with `make test`
- [x] I ensured my PR is ready for review with `make reviewable`

thorbenbelow commented:
Hi @Skarlso @knelasevero, is there anything I'm missing or that I can do to further assist you with this PR?

Skarlso commented on pkg/provider/passbolt/passbolt.go:
```diff
@@ -0,0 +1,298 @@
+}
+
+func (provider *ProviderPassbolt) PushSecret(_ context.Context, _ *corev1.Secret, _ esv1beta1.PushSecretData) error {
+	return nil
```
Please return an error for unimplemented methods.

Skarlso commented on pkg/provider/passbolt/passbolt.go:
```diff
@@ -0,0 +1,298 @@
+		return nil, err
+	}
+
+	nameRegexp, err := regexp.Compile(ref.Name.RegExp)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, resource := range resources {
+		if !nameRegexp.MatchString(resource.Name) {
+			continue
+		}
```
This makes the RegExp parameter mandatory since it won't return anything unless you define something like "*.".

But in your API definition the Regex is an optional field.

thorbenbelow replied:
I added an error for the case of a missing `RegExp` parameter.

Please let me know if this is insufficient or if you prefer another approach.

Skarlso replied:
Fine by me.

Skarlso commented on pkg/provider/passbolt/passbolt.go:
```diff
@@ -0,0 +1,298 @@
+			continue
+		}
+
+		secret, err := getPassboltSecret(ctx, provider.client, resource.ID)
```
Put this method on the provider so you don't have to pass in and copy the client.

Skarlso commented on pkg/provider/passbolt/passbolt.go:
```diff
@@ -0,0 +1,298 @@
+		if err != nil {
+			return nil, err
+		}
+		marshaled, err := json.Marshal(secret)
```
use utils.Marshal.

Skarlso reviewed (changes requested):
A few remarks but overall a good approach!

thorbenbelow commented:
Hi @Skarlso, thank you very much for the feedback.
I added the requested changes and updated the branch.
Let me know if there is anything else I should change.
Do you have a policy regarding squashing commits?
